File: algorithms/greedy/ValidPalindrome.py
"""
Valid Palindrome II
https://leetcode.com/problems/valid-palindrome-ii/

Given a string s,
return true if the s can be palindrome after deleting at most one character from it.

Example 1:
    Input: s = "aba"
    Output: true

Example 2:
    Input: s = "abca"
    Output: true
    Explanation: You could delete the character 'c'.

Example 3:
    Input: s = "abc"
    Output: false

Constraints:
    * 1 <= s.length <= 10^5
    * s consists of lowercase English letters.
"""


class ValidPalindrome:

    # Algorithm(s) Used: Two Pointers
    # Time Complexity: O(n)
    # Space Compexlity: O(1)
    def is_palindrome(self, s: str, left: int, right: int) -> bool:
        while left < right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True

    # Algorithm(s) Used: Greedy, Two Pointers
    # Time Complexity: O(n)
    # Space Compexlity: O(1)
    def validPalindromeI(self, s: str) -> bool:
        l, r = 0, len(s) - 1

        while l < r:
            if s[l] != s[r]:  # Local Choice - First Mismatch (Allowed)
                return (      # Greedy Choice - Second Mismatch (Not Allowed)
                    self.is_palindrome(s, l + 1, r) or
                    self.is_palindrome(s, l, r - 1)
                )

            # Local Optimal Solution - Update Pointers
            l += 1
            r -= 1

        return True


if __name__ == "__main__":
    Solution = ValidPalindrome()

    test_cases = [
        # (s, expected)
        ("aba", True),
        ("abca", True),
        ("abc", False),
        ("a", True),
        ("aa", True),
        ("ab", True),
        ("deeee", True),
        ("racecar", True),
        ("radar", True),
        ("abcdef", False),
    ]

    funcs = [
        Solution.validPalindromeI,
    ]

    for func in funcs:
        for s, expected in test_cases:
            result = func(s)
            assert result == expected

File: algorithms/greedy/ValidPalindrome.cpp
/*
Valid Palindrome II
https://leetcode.com/problems/valid-palindrome-ii/

Given a string s,
return true if the s can be palindrome after deleting at most one character from it.

Example 1:
    Input: s = "aba"
    Output: true

Example 2:
    Input: s = "abca"
    Output: true
    Explanation: You could delete the character 'c'.

Example 3:
    Input: s = "abc"
    Output: false

Constraints:
    * 1 <= s.length <= 10^5
    * s consists of lowercase English letters.
*/

#include <cassert>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class ValidPalindrome {
private:

    // Algorithm(s) Used: Two Pointers
    // Time Complexity: O(n)
    // Space Compexlity: O(1)
    bool isPalindrome(string s, int left, int right) {
        while (left < right) {
            if (s[left] != s[right]) {
                return false;
            }
            left += 1;
            right -= 1;
        }

        return true;
    }

public:

    // Algorithm(s) Used: Greedy, Two Pointers
    // Time Complexity: O(n)
    // Space Compexlity: O(1)
    bool validPalindromeI(string s) {
        int left = 0;
        int right = s.size() - 1;

        while (left < right) {
            // Local Choice - First Mismatch (Allowed)
            if (s[left] != s[right]) {
                // Greedy Choice - Second Mismatch (Not Allowed)
                return (
                    isPalindrome(s, left + 1, right) || isPalindrome(s, left, right - 1)
                );
            }

            // Local Optimal Solution - Update Pointers
            left += 1;
            right -= 1;
        }

        return true;
        
    }
};

int main() {
    ValidPalindrome Solution;

    // (string s, expected)
    vector<tuple<string, bool>> test_cases = {
        {"aba", true},
        {"abca", true},
        {"abc", false},
        {"a", true},
        {"aa", true},
        {"ab", true},
        {"deeee", true},
        {"racecar", true},
        {"radar", true},
        {"abcdef", false},
    };

    vector<bool (ValidPalindrome::*)(string)> funcs = {
        &ValidPalindrome::validPalindromeI,
    };

    for (auto func : funcs) {
        for (auto& tc : test_cases) {
            string s;
            bool expected;
            tie(s, expected) = tc;

            bool result = (Solution.*func)(s);
            assert(result == expected);
        }
    }

    return 0;
}
